[cweijin] iP

cweijin opened the pull request:
### Mylo

> "You can have it all. Just not all at once." - Oprah Winfrey ([source](https://ca.indeed.com/career-advice/career-development/quotes-on-time-management))

**Mylo helps you declutter your mind by keeping track of everything you need to do. It’s:**

- text-based
- intuitive and easy to use
- ~~FAST~~ _LIGHTNING FAST_—get things done in no time!

Getting started is simple:

1. Download it from [here](https://github.com/cweijin/ip/releases).
2. Double-click to launch.
3. Add your tasks.
4. Let Mylo handle the rest! 😉

And the best part? It’s completely **FREE**!

Features:

- [x] Task management
- [x] Deadlines
- [ ] Smart reminders (coming soon)

---
> [!NOTE]
> If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        System.out.println("Hello World");
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

chongwj13 committed:
Chatbot renamed to Mylo. Modified the greeting and goodbye messages.

chongwj13 committed:
Implemented echo-ing. Added static method separator() for ease of use.

chongwj13 committed:
Introduced TaskList and TaskModel with methods as per required for level 2. Improve the flow control in Mylo.java using switch-case.

chongwj13 committed:
Now able to mark and unmark the completion status of a task. Introduced methods such as getStatusIndicator, markTaskAsDone, markTaskAsUndone.

chongwj13 committed:
Implemented 3 task types: Todo, Deadline, Event. Uses factory method in Task class to decide which subtype to return. Modified the message printed. \nNew Command: 1. todo ... \n2. event ... /from ... /to ... \n3. deadline ... /by ...

cweijin committed:
Implemented InsufficientInfoException and NoSuchCommandException. Errors such as invalid input and insufficient information are being handled.

cweijin committed:
Implemented delete command. delete <index> will delete the task at the index and prints deleted message if successful.

cweijin committed:
Implement automated text UI testing

Automated Text UI Testing is required to semi-automate the testing of Mylo.

input.txt when empty causes Mylo to throw exception since there is no next line.

Returning goodbye message is suitable when there is no next line.

Let's,
* modify Mylo.java to have input initiated to bye
* update EXPECTED.TXT to include goodbye message

cweijin committed:
Implement automatic task save

Task list is temporary and lost after program terminates.

Automatic task save allows users to have their task list retained for next program execution.

Let's,
* create Storage class to handle the logic of save, rewrite and load local storage file
* create some exception classes to handle corrupted file format.

cweijin committed:
Merge bran-Level-7 into master branch

cweijin committed:
Implement dates and time

Date and time is stored as String.

LocalDateTime provides more meaningful methods such as to compare between two DateTime.

Let's,
* change the type for date to LocalDateTime
* add command to print deadlines/events on specific date
* add formatter and helper function in formatting date

cweijin committed:
Merge branch-Level-8 into master branch

keithxun commented on src/main/java/Main.java:
```diff
@@ -0,0 +1,137 @@
+    /**
+     * Initializes the required objects and loads up the data from the storage file.
+     */
+    public void init() throws FileCorruptedException, StorageOperationException {
```
I like that you handled file exceptions

keithxun commented on src/main/java/Main.java:
```diff
@@ -0,0 +1,137 @@
+    /**
+     * Reads the user command and executes it, until the user issues the exit command.
+     */
+    public void runCommandLoopUntilExitCommand() {
```
Long function name but I think it is suitable

keithxun commented on src/main/java/data/InsufficientInfoException.java:
```diff
@@ -0,0 +1,21 @@
+            case TODO -> "Oops. task.Task title is required. Please try again with: todo <task.Task Title>.";
+            case EVENT ->
+                    "Oops. Seems like we are missing some information. Please try again with: event <task.Task Title> /from <dd-mm-yyyy hhmm> /to <dd-mm-yyyy hhmm>";
+            case DEADLINE ->
```
Do check that your switch statement follows coding convention (switch and case on the same indentation)

keithxun commented on src/main/java/task/Event.java:
```diff
@@ -0,0 +1,40 @@
+    public Event(String title, String startDate, String endDate, boolean isDone) throws IllegalValueException {
+        super(title, isDone);
+        this.startDate = HelperFunctions.stringToDateTime(startDate);
+        this.endDate = HelperFunctions.stringToDateTime(endDate);
```
Maybe you can consider handling time too since the package has those functionalities too

keithxun commented on src/main/java/utils/helpers/HelperFunctions.java:
```diff
@@ -0,0 +1,34 @@
+
+import java.time.LocalDateTime;
+
+public class HelperFunctions {
```
I like how you have a file to store helper functions. Are they the same as components?

keithxun reviewed (commented):
lgtm

cweijin committed:
Refactor the code to make it more OOP

Code not closely adhere to OOP.

Code is not well organized without OOP.

Abstracting out closely related code as classes give a more organized code and is good for future increment.

Let's,
* abstract out closely related code into classes
* create Command, Parser, Ui, Storage and Task classes
* adhere to OOP in future increments

cweijin committed:
Merge branch-A-MoreOOP to master

cweijin committed:
Merge add-gradle-support to master

cweijin committed:
Use gradle to automate some of the build tasks

cweijin committed:
Merge branch-A-Gradle to master

cweijin committed:
Add JUnit tests

Methods in classes are not tested before commit.

JUnit helps test methods to detect any bugs when codes are modified.

Let's,
* merge build-gradle branch
* set up JUnit test for non trivial methods

cweijin committed:
Merge branch-A-JUnit to master

cweijin committed:
Add JavaDoc comments

Classes and methods do not have JavaDoc comments to explain their uses.

JavaDoc comments are required to provide code readibility and ease future code maintenance.

Let's add JavaDoc header comments to all non-private classes/methods.

cweijin committed:
Merge A-Javadoc to master

cweijin committed:
Comply with coding standard

cweijin committed:
Merge branch-A-CodingStandard to master

cweijin committed:
Implement find feature

User cannot search for tasks by keyword

It would be much more convenient if users are allowed to search using keyword.

Let's,
* introduce FindCommand class
* update Parser to identify find command
* implement tasksWithKeyword method

cweijin committed:
Merge branch-Level-9 to master

cweijin committed:
Use checkstyle to detect coding style violations

cweijin committed:
Merge branch-A-CheckStyle to master

cweijin committed:
Add Gui

Mylo is TUI-based.

A GUI can provide better user experience.

Let's,
* implement GUI for Mylo using JavaFX
* use FXML to separate GUI design from controller
* modify Tui class to only prints input and messages

cweijin committed:
Merge branch-Level-10 to master

cweijin committed:
Use assert feature

Assumptions are made without using assert feature to document them.

Assertion helps document importatnt assumptions that should hold at various points in the code.

Let's,
* use assert feature when assumptions are made
* enable assertions in gradle

cweijin committed:
Merge pull request #2 from cweijin/branch-A-Assertions

Use assert feature

cweijin committed:
Improve code quality

Small part of the initial code does not follow code quality.

Good code quality is important for future code maintenance.

Let's,
* examine the code and refactor to improve code quality
* follow closely the 'Code Quality' topics

cweijin committed:
Merge pull request #3 from cweijin/branch-A-CodeQuality

Improve code quality

cweijin committed:
Set up continuous integration

No error detection workflow when updating repo.

CI is important to detect errors in new update.

Let's create a gradle.yml workflow for CI

cweijin committed:
Using streams

Operations that involve filter are done using loops.

Java provides a more efficient way by using streams.

Let's change the code to use streams together with its filter
and map methods to make operations more efficient.

cweijin committed:
Merge pull request #5 from cweijin/branch-A-Streams

Using streams

cweijin committed:
Set up continuous integration

No error detection workflow when updating repo.

CI is important to detect errors in new update.

Let's create a gradle.yml workflow for CI

cweijin committed:
Merge pull request #6 from cweijin/branch-A-CI

Set up continuous integration

cweijin committed:
Detect duplicates

Duplicates are not being handled.

Handling duplication can boost user experience as users might not
be aware of the duplication causing mess in task list.

Let's:
* create DuplicatedTaskException
* handle duplication error in addTask()
* override equals method in Task to compare between them

cweijin committed:
Detect duplicates

Duplicates are not being handled.

Handling duplication can boost user experience as users might not
be aware of the duplication causing mess in task list.

Let's:
* create DuplicatedTaskException
* handle duplication error in addTask()
* override equals method in Task to compare between them

cweijin committed:
Merge pull request #7 from cweijin/branch-C-DetectDuplicates

Detect duplicates

cweijin committed:
Improve the Gui

Gui is not polished.

A better Gui gives better user experience.

Let's:
* Differentiate the layout of user input and Mylo's response
* Differentiate error message and normal response from Mylo
* Clip the user profile photo in circle
* Change the send button to icon
* Design the message bubble

cweijin committed:
Merge pull request #8 from cweijin/branch-A-BetterGui

Improve the Gui

cweijin committed:
Junit tests

JUnit tests are not enough to fully test the classes.

Automated tests like JUnit test helps to detect bugs early.

Let's complete the JUnit test to test TaskList and Event class.

cweijin committed:
Bug fix: git actions check

cweijin committed:
Bug fix: instantiate tasklist when each test is called

cweijin committed:
Bug fix: show error message when test failed

cweijin committed:
Merge pull request #9 from cweijin/branch-A-MoreTesting

Junit tests

cweijin committed:
Add ai usage record

cweijin committed:
Merge pull request #10 from cweijin/branch-A-AiAssisted

Add ai usage record

cweijin committed:
Bug fix: io exception

cweijin committed:
Set up product website

cweijin committed:
Merge pull request #11 from cweijin/branch-UG

Set up product website

cweijin committed:
Refine user guide

cweijin committed:
Optimize import

cweijin committed:
Merge pull request #12 from cweijin/branch-UG

Refine user guide

cweijin committed:
Refine readme format
